### ISSUE TITLE
Feature request: unsound-assignment rule

### ISSUE THREAD

**Issue by jelle-openai** [astral-sh/ty#4289]
ty recently added strict rules like https://docs.astral.sh/ty/reference/rules/#unsound-yield, which are good for making sure there's no type unsafety in code. But these rules miss some adjacent unsoundness:

```python
from typing import Generator, Iterable, Any

def f():
    return "hi"

def g() -> Generator[int]:
    var: int = f() # no error on either line even with all settings enabled
    yield var
```

I propose a new rule unsound-assignment, off by default (like the yield and return rules) that would flag this assignment.

**Comment by carljm** [comment]
Makes sense. Not marking this for stable release (and not sure if it should be tagged `strict` or not, per the current description of that label) since it is not required for mypy-strict parity.

The mypy-strict features that do fire on this code example are `no-untyped-def` and `no-untyped-call`. But those should be distinct issues.
